ygot/diff.go

```diff
@@ -210,5 +235,155 @@ func findSetLeaves(s GoStruct) (map[*pathSpec]interface{}, error) {
 		return nil, fmt.Errorf("error from ForEachDataField iteration: %v", errs)
 	}
 
-	return out, nil
+	uout := map[*pathSpec]interface{}{}
+	// Deduplicate the list, since the iteration function will be called
+	// multiple times for path tags that have >1 element.
+	for ok, ov := range out {
+		var skip bool
+		for uk := range uout {
+			if ok.Equal(uk) {
+				// This is a duplicate path, so we do not need to append it to the list.
+				skip = true
+			}
+		}
+		if !skip {
+			uout[ok] = ov
+		}
+	}
+
+	return uout, nil
+}
+
+// togNMIValue returns the GoStruct field v as a gNMI TypedValue message. It
+// is a wrapper around the openconfig/gnmi/value FromScalar library function with
+// handling for GoStruct specific types, particularly:
+//  - Enumerated values (YANG enum or identityref leaves) which are returned as
+//    their string value, per the gNMI spec.
+//  - Union values - which are extracted from the interface type that is used
+//    to represent them.
+func togNMIValue(v interface{}) (*gnmipb.TypedValue, error) {
+	val := reflect.ValueOf(v)
+	switch {
+	case val.Type().Kind() == reflect.Int64:
+		if _, ok := v.(GoEnum); ok {
+			estr, _, err := enumFieldToString(reflect.ValueOf(v), false)
+			if err != nil {
+				return nil, fmt.Errorf("cannot convert enum field to string: %v", err)
+			}
+
+			return &gnmipb.TypedValue{
+				Value: &gnmipb.TypedValue_StringVal{estr},
+			}, nil
+		}
+		// A non-enum implementing int64 is an invalid type, since fields should
+		// be pointers.
+		return nil, fmt.Errorf("invalid enum type")
+	case util.IsValueStructPtr(val):
+		var err error
+		if v, err = unionInterfaceValue(val, false); err != nil {
+			return nil, fmt.Errorf("cannot resolve union field value: %v", err)
+		}
+	case util.IsValuePtr(val):
+		v = val.Elem().Interface()
+	case val.Type().Name() == EmptyTypeName:
+		v = val.Bool()
+	case val.Type().Name() == BinaryTypeName:
+		v = val.Bytes()
+	}
+
+	return value.FromScalar(v)
+}
+
+// appendUpdate adds an update to the supplied gNMI Notification message corresponding
+// to the path and value supplied.
+func appendUpdate(n *gnmipb.Notification, path *pathSpec, val interface{}) error {
+	v, err := togNMIValue(val)
+	if err != nil {
+		s := pretty.Sprint(path)
```

Comment:
I wonder - would this and the following lines look better as a method on the pathSpec type?

Maybe we can even have it implement fmt.Stringer and then just use %v here in the fnt.Errorf() call.

Reply:
SGTM! Implemented `String()` with a `pathSpec` receiver.